experimental/Schemes/WeilDivisor.jl

```diff
@@ -0,0 +1,352 @@
+
++(D::WeilDivisor, I::IdealSheaf) = D + WeilDivisor(I)
+
+function ==(D::WeilDivisor, E::WeilDivisor) 
+  keys(coefficient_dict(D)) == keys(coefficient_dict(E)) || return false
+  for I in keys(coefficient_dict(D))
+    if haskey(coefficient_dict(E), I)
+      D[I] == E[I] || return false
+    else
+      iszero(D[I]) || return false
+    end
+  end
+  for I in keys(coefficient_dict(E))
+    !(I in keys(coefficient_dict(D))) && !(iszero(E[I])) && return false
+  end
+  return true
+end
+
+#function intersection(D::T, E::T) where {T<:WeilDivisor}
+#  X = scheme(D)
+#  X == scheme(E) || error("divisors do not live on the same scheme")
+#  R = coefficient_ring(D)
+#  R == coefficient_ring(E) || error("divisors do not have the same coefficient ring")
+#  # prepare a copy of the divisors
+#  D_copy = WeilDivisor(X, R)
+#  E_copy = WeilDivisor(X, R)
+#  # check whether a common refinement of the covering is necessary
+#  CD = covering(D)
+#  CE = covering(E)
+#  if CD != CE
+#    CC, f, g = common_refinement(X, CD, CE)
+#    D_copy = pullback(f, D)
+#    E_copy = pullback(g, E)
+#  else
+#    D_copy = D
+#    E_copy = E
+#  end
+#  # TODO: Work out the intersection
+#end
+
+@Markdown.doc """
+    in_linear_system(f::VarietyFunctionFieldElem, D::WeilDivisor; check::Bool=true)
+
+Returns `true` if the rational function `f` is in the linear system ``|D|``
+and `false` otherwise.
+"""
+function in_linear_system(f::VarietyFunctionFieldElem, D::WeilDivisor; check::Bool=true)
+  X = scheme(D) 
+  X === variety(parent(f)) || error("schemes not compatible")
+  C = default_covering(X)
+  for I in components(D)
+    order_on_divisor(f, I, check=check) >= -D[I] || return false
+  end
+  for U in patches(C)
+    # we have to check that f[U] has no poles outside D[U]
+  end
+  return true
+end
+
+@Markdown.doc """
+    LinearSystem
+
+A linear system of a Weil divisor `D` on a variety `X`, 
+generated by rational functions ``f₁,…,fᵣ ∈ K(X)``.
+"""
+@attributes mutable struct LinearSystem{DivisorType<:WeilDivisor}
+  D::DivisorType
+  f::Vector{<:VarietyFunctionFieldElem}
+
+  function LinearSystem(f::Vector, D::WeilDivisor; check::Bool=true)
+    length(f) == 0 && return new{typeof(D)}(D, Vector{VarietyFunctionFieldElem}())
+    KK = parent(f[1])
+    all(g -> (parent(g) === KK), f[2:end]) || error("elements must have the same parent")
+    X = scheme(D)
+    X === variety(KK) || error("input not compatible")
+
+    if check
+      all(g->in_linear_system(g, D), f) || error("element not in linear system")
+    end
+    f = Vector{VarietyFunctionFieldElem}(f)
+    return new{typeof(D)}(D, f)
+  end
+end
+
+### essential getters 
+@Markdown.doc """
+    weil_divisor(L::LinearSystem)
+
+Return the divisor `D` of the linear system `L = |D|`.
+"""
+function weil_divisor(L::LinearSystem) 
+  return L.D
+end
+gens(L::LinearSystem) = L.f
+ngens(L::LinearSystem) = length(L.f)
+@Markdown.doc """
+    variety(L::LinearSystem)
+
+Return the variety on which `L` is defined.
```

Comment:
scheme

Reply:
But didn't we want to make a distinction here? We do not allow linear systems to be defined on arbitrary schemes. These were varieties. But not with their own type, because we decided to store these in the attributes for the time being.

Reply:
Hartshorne: "An abstract variety is an integral separated scheme of finite type over an algebraically closed field k."
So excludes almost anything in the schemes because we are not algebraically closed. 
We want a noetherian integral separated scheme which is normal (or at least regular in codimension one).